### PR TITLE
Massive wire render optimization

### PR DESCRIPTION
**Pull request by thegrb93**
~~Not tested completely, but passes basic tests.~~

### CODE REVIEW

**Comment by thegrb93** [comment]
Added a little more optimization. Should fix https://github.com/wiremod/wire/issues/921 btw

**Comment by AbigailBuccaneer** [comment]
Could I perhaps talk you into doing #969 by entirely replacing the beam vars library with something nicer from scratch? (This was on my to do list but this pull request is already most of the way there!)

**Comment by thegrb93** [comment]
Ok I think it's working even faster now.

**Comment by Divran** [comment]
So are beam vars even used at all any more after this? Also, did you remember to keep support for blinking wires? The wire tool(s) use blinking wires to show the user where inputs are wired to.

**Comment by thegrb93** [comment]
Yeah blinking still works. The wire tool seems to not update the wire when adding links until the wire is finished though, but that's not an issue with this pr.

**Comment by Divran** [comment]
yeah the wire tool is all clientside, and doesn't send the info until you're done

**Comment by thegrb93** [comment]
```
[ERROR] lua/includes/extensions/net.lua:74: net.WriteColor: color expected, got table
  1. assert - [C]:-1
   2. WriteColor - lua/includes/extensions/net.lua:74
    3. TransmitPath - addons/wire/lua/wire/wire_paths.lua:76
     4. unknown - addons/wire/lua/wire/wire_paths.lua:90
```
Darn, gotta make color copy.

**Comment by thegrb93** [comment]
Should be good now.

**Comment by thegrb93** [comment]
Yeah working fine now. Running on my server with 20 players.

File: lua/wire/wire_paths.lua [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		local num_points = net.ReadUInt(15)
+		for i = 1, num_points do
+			path.Path[i] = { Entity = net.ReadEntity(), Pos = net.ReadVector() }
+		end
+
+		if path.Entity.WirePaths == nil then path.Entity.WirePaths = {} end
+		path.Entity.WirePaths[path.Name] = path
+
+	end)
+
```
Comment (AbigailBuccaneer) [inline lua/wire/wire_paths.lua]: Trailing whitespace.

File: lua/wire/client/cl_wirelib.lua [hunk 2]
```diff
@@ -34,181 +34,79 @@ end
...
+			if lastrender ~= t then
+				local w, f = math.modf(t*WIRE_BLINKS_PER_SECOND)
+				shouldblink = f < 0.5
+				scroll = t*WIRE_SCROLL_SPEED
+				lastrender = t
+			end
 
-					local width = ent:GetNetworkedBeamFloat(net_name .. "_width")
-					if width > 0 then
+			local blink = shouldblink and ent:GetNetworkedBeamString("BlinkWire")
```
Comment (AbigailBuccaneer) [inline lua/wire/client/cl_wirelib.lua]: `"BlinkWire"` is now the only place in Wiremod that uses `{Get,Set}NetworkedBeam*`. Can you kill it off (by just replacing it with `{Get,Set}NWString`) and then remove `beam_netvars.lua`?

File: lua/wire/wire_paths.lua [hunk 3]
```diff
@@ -0,0 +1,116 @@
+-- wire_paths.lua
+--
+-- This file is implements syncing of wire paths, which are the visual
```
Comment (AbigailBuccaneer) [inline lua/wire/wire_paths.lua]: (typo, my fault)

**Comment by thegrb93** [comment]
K

**Comment by thegrb93** [comment]
Forgot to remove it from wire_load. Testing now.

**Comment by thegrb93** [comment]
Works

**Comment by thegrb93** [comment]
Turns out there are some mods that use networkedbeams and rely on wiremod having it :/

**Comment by AbigailBuccaneer** [comment]
@thegrb93 the BeamVars library is written to handle multiple addons both shipping it - we should give them a heads up that we're removing it, but I don't like the idea of keeping it because other addons expect us to.

We could ship a version with `ThisBeamNetVarsVersion = -1` that prints error messages (blaming the addon that called the functions) and redirecting `{Get,Set}NWBeam*` to `{Get,Set}NW*`?

**Comment by AbigailBuccaneer** [comment]
I've just submitted an issue for ACF-Missiles, but I can't find (via GitHub's code search) anywhere in [nrlulz/ACF](https://github.com/nrlulz/ACF) that uses it.

File: lua/wire/wire_paths.lua [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	if ent:IsValid() and ent.Inputs then
+		for name, input in pairs(ent.Inputs) do
+			if input.Src then
+				WireLib.Paths.Add(path, ply)
+			end
+		end
+	end
+end)
+
+local function TransmitPath(input)
```
Comment (CaptainPRICE) [inline lua/wire/wire_paths.lua]: Pick a better name for `input` argument.

**Comment by thegrb93** [comment]
Okay I reverted that last commit. Good to go.

File: lua/wire/server/wirelib.lua [hunk 5]
```diff
@@ -578,28 +577,18 @@ function WireLib.Link_Start(idx, ent, pos, iname, material, color, width)
 
 	local input = ent.Inputs[iname]
```
Comment (thegrb93) [inline lua/wire/server/wirelib.lua]: I don't think anyone is going to confuse it with the input library.

**Comment by oldmud0** [comment]
Wow, this is big. What kind of performance improvements are we seeing from a pragmatic standpoint? FPS? latency?

**Comment by thegrb93** [comment]
Less network usage when pasting a dupe, and less client cpu usage per wire entity per frame.

File: lua/wire/server/wirelib.lua (hunk 5 above)
Comment (AbigailBuccaneer) [inline lua/wire/server/wirelib.lua]: Wiremod has used this name for an incredibly long time, and if we were to change it now we'd at the very least have to have a conversation about preserving backwards compatibility. This is a discussion for sometime and someplace else.

**Comment by thegrb93** [comment]
bump

File: lua/wire/client/cl_wirelib.lua [hunk 6]
```diff
@@ -1,7 +1,7 @@
 local WIRE_SCROLL_SPEED = 0.5
 local WIRE_BLINKS_PER_SECOND = 2
 local CurPathEnt = {}
-local Wire_DisableWireRender = 2 --bug with mode 0 and gmod2007beta
+local Wire_DisableWireRender = 0 --bug with mode 0 and gmod2007beta
```
Comment (TomyLobo) [inline lua/wire/client/cl_wirelib.lua]: Can you review the comment in this line and decide whether to keep it or not?

File: lua/wire/server/wirelib.lua (hunk 5 above)
Comment (TomyLobo) [inline lua/wire/server/wirelib.lua]: @AbigailBuccaneer this is a local variable, your argument is invalid.
luacheck is going to complain about shadowing a global, too.

Comment (AbigailBuccaneer) [inline lua/wire/server/wirelib.lua]: Oh, whoops, I misread the comment as complaining about the name of the `Inputs` field, not the variable. Finding time to work on Wiremod means sleep deprivation usually :wink:

**Comment by thegrb93** [comment]
k

File: lua/wire/server/wirelib.lua (hunk 5 above)
Comment (thegrb93) [inline lua/wire/server/wirelib.lua]: WireLib.lua also contains local input as well.

**Comment by thegrb93** [comment]
Bump again.

**Comment by Divran** [comment]
Looks good (I haven't tested it ingame though). Merge?

**Comment by thegrb93** [comment]
I've had no problems with it.

**Comment by Python1320** [comment]
```
Player [5][Not Mike] ERR: lua/wire/wire_paths.lua:41: attempt to index field 'WirePaths' (a nil value)
```
Null entity check missing?

File: lua/wire/wire_paths.lua [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		path.Material = net.ReadString()
+		path.Color = net.ReadColor()
+
+		local num_points = net.ReadUInt(15)
+		for i = 1, num_points do
+			path.Path[i] = { Entity = net.ReadEntity(), Pos = net.ReadVector() }
+		end
+
+		if path.Entity.WirePaths == nil then path.Entity.WirePaths = {} end
+		path.Entity.WirePaths[path.Name] = path
```
Comment (thegrb93) [inline lua/wire/wire_paths.lua]: I see. Looks like that callback needs IsValid check too. I'll add it once I get home. Would Entity:IsValid() return false in the outside pvs case too?

**Comment by thegrb93** [comment]
Keeping beamvars because WireExtras and ACF rely on it still.

**Comment by Anticept** [comment]
Since this was approved, here we go!

**Pull request merged**